### DISCUSSION TITLE
Is there a way to specify `ToUpper` or `ToLower` when formatting data in an interpolated `string`?

### DISCUSSION THREAD

**Discussion by tacosontitan** [dotnet/runtime#98012] [Q&A] [answered]
I want to print today's date in the format `d MMM yyyy` and it's easy to do with interpolation:

```cs
var output = $"Today is {DateTime.Now:d MMM yyyy}.";
Console.WriteLine(output);
```

Assuming today's date is January 1st, 2024, the expected output of the above snippet is:

> Today is 1 Jan 2024.

However, I have an added requirement that I want to display this in uppercase, but that currently (to my knowledge) requires manual invocations of `ToString` and `ToUpper` to achieve:

```cs
var output = $"Today is {DateTime.Now.ToString("d MMM yyyy".ToUpper()}.";
Console.WriteLine(output);
```

This achieves the desired effect, but I can't help but feel like this is something that would have support without those manual invocations. Is there a way to specify `ToUpper` or `ToLower` when formatting data in an interpolated `string`?

> [!IMPORTANT]
> I'm not looking to format the entire `output` variable, simply the date within it; this makes the desired output:  
> > Today is 1 JAN 2024.

**Comment by danmoseley** [comment]
Are you trying to avoid allocations? Ie this path is very hot.

**Reply by tacosontitan** [reply]
In this particular case, yes, but I've had this exact desire on more than a handful of occasions some hot, some not.

**Comment by Joe4evr** [comment]
Looks like you could create an instance of `DateTimeFormatInfo` where you set `AbbreviatedMonthNames` to the desired values, and pass that along with the format string. Maybe look at some documentation for specifics because it looks like you must set an array of exactly *13* elements (with index 0 `null`?)

**Reply by Clockwork-Muse** [reply]
Given that the question is implying an abuse of culture data for interchange purposes (even if OP isn't the originator of the requirement), this would normally be the way to go.

@tacosontitan - 
For an explicit format scenario like this, you should be supplying a fixed culture anyways (normally this would be `InvariantCulture`).  That actually makes things a little more complicated for formatting strings, though....
Also, in the interest of thoroughness, you should be getting the current time from a `TimeProvider` instance for testability reasons, if your real code doesn't do that already.

**Reply by Joe4evr** [reply]
I'm not seeing an implication that this question is for interchange, only for display. I would seriously hope that systems which take in timestamps in their payloads are parsing them from a more standardized format like ISO 8601 (`"O"` in .NET), otherwise I very much question the quality.

**Reply by Clockwork-Muse** [reply]
.... Sure, but if it's for display it's bug-for-bug compatible behavior.

There shouldn't normally be a reason to customize the output this way, since this ignores any potential user format settings.
In the rare case that this may be some sort of spreadsheet program, then I really hope they're doing some sort of culture merging for driving display values with custom formats ....

**Reply by Joe4evr** [reply]
> There shouldn't normally be a reason to customize the output this way, since this ignores any potential user format settings.

Here's a first-hand experience reason: As of the start of this year, my workplace mandated *the same format* for any date placed in documents sent to customers in order to be unambiguous all around the world. (I'm personally less affected because I don't interact with the company's customers, but it was something everyone had to be notified of.)

OP's "added requirement" is likely similar.

**Reply by tacosontitan** [reply]
While I appreciate the feedback to be culture cognizant, this is a case where I've done what I can to sway the decision and just have to implement it at this point, and my current solution works fine, it just seems redundant.

That being said, my question is likely too targeted on `DateTime` consequently causing people to treat it as an X/Y problem. The real question here is if casing can be specified when formatting data in an interpolated string.

**Reply by Joe4evr** [reply, marked as answer]
> The real question here is if casing can be specified when formatting data in an interpolated string.

That's a no, sadly. My other answer would work for anything `ISpanFormattable` if you're willing to change from interpolation to like a `StringBuilder` or something, or just live with the intermediate allocation.

**Comment by Joe4evr** [comment]
Another option (which is a bit more involved since I believe it doesn't work nicely with string interpolation) would be:
* Create a `Span<char>` buffer that fits your format (looks like 11 chars should be sufficient, so this could even be `stackalloc`ed)
* Fill the buffer using `TryFormat`
* Use `ToUpper()` on the span which I believe does the conversion in-place

Then just pick whatever method you like to copy the contents of the buffer into your final string (the real minimal-alloc way is `String.Create`, but ultimately it's your choice).

**Comment by TrayanZapryanov** [comment]
@tacosontitan check [my question](https://github.com/dotnet/runtime/discussions/97795)

You can use the same approach to build ToUpper for every object and to use it in interpolated strings.
